rcljava/src/main/cpp/org_ros2_rcljava_action_ActionServerImpl.cpp

```diff
@@ -0,0 +1,386 @@
+    rcl_reset_error();
+    rcljava_throw_rclexception(env, ret, msg);
+    return NULL;
+  }
+
+  jbooleanArray result = env->NewBooleanArray(4);
+  jboolean temp_result[4] = {
+    is_goal_request_ready,
+    is_cancel_request_ready,
+    is_result_request_ready,
+    is_goal_expired
+  };
+  env->SetBooleanArrayRegion(result, 0, 4, temp_result);
+  return result;
+}
+
+JNIEXPORT void JNICALL
+Java_org_ros2_rcljava_action_ActionServerImpl_nativeDispose(
+  JNIEnv * env, jclass, jlong node_handle, jlong action_server_handle)
+{
+  if (action_server_handle == 0) {
+    // everything is ok, already destroyed
+    return;
+  }
+
+  if (node_handle == 0) {
+    // TODO(jacobperron): throw exception
+    return;
+  }
+
+  rcl_node_t * node = reinterpret_cast<rcl_node_t *>(node_handle);
+
+  assert(node != NULL);
+
+  rcl_action_server_t * action_server = reinterpret_cast<rcl_action_server_t *>(
+    action_server_handle);
+
+  assert(action_server != NULL);
+
+  rcl_ret_t ret = rcl_action_server_fini(action_server, node);
+
+  if (ret != RCL_RET_OK) {
+    std::string msg = "Failed to destroy action server: " +
+      std::string(rcl_get_error_string().str);
+    rcl_reset_error();
+    rcljava_throw_rclexception(env, ret, msg);
+  }
+}
+
+JNIEXPORT jlong JNICALL
+Java_org_ros2_rcljava_action_ActionServerImpl_nativeCreateActionServer(
+  JNIEnv * env,
+  jobject,
+  jlong node_handle,
+  jlong clock_handle,
+  jclass jaction_class,
+  jstring jaction_name)
+{
+  jmethodID mid = env->GetStaticMethodID(jaction_class, "getActionTypeSupport", "()J");
+  assert(mid != NULL);
+
+  jlong jts = env->CallStaticLongMethod(jaction_class, mid);
+  assert(jts != 0);
+
+  const char * action_name_tmp = env->GetStringUTFChars(jaction_name, 0);
+
+  std::string action_name(action_name_tmp);
+  env->ReleaseStringUTFChars(jaction_name, action_name_tmp);
+
+  rcl_node_t * node = reinterpret_cast<rcl_node_t *>(node_handle);
+  rcl_clock_t * clock = reinterpret_cast<rcl_clock_t *>(clock_handle);
+
+  rosidl_action_type_support_t * ts = reinterpret_cast<rosidl_action_type_support_t *>(jts);
+
+  rcl_action_server_t * action_server = static_cast<rcl_action_server_t *>(
+    malloc(sizeof(rcl_action_server_t)));
+  *action_server = rcl_action_get_zero_initialized_server();
+  rcl_action_server_options_t action_server_ops = rcl_action_server_get_default_options();
+
+  rcl_ret_t ret = rcl_action_server_init(
+    action_server, node, clock, ts, action_name.c_str(), &action_server_ops);
+  // env->ReleaseStringUTFChars(jaction_name, action_name);
+
+  if (ret != RCL_RET_OK) {
+    std::string msg = "Failed to create action server: " + std::string(rcl_get_error_string().str);
+    rcl_reset_error();
+    rcljava_throw_rclexception(env, ret, msg);
+    free(action_server);
+    return 0;
+  }
+
+  jlong jaction_server = reinterpret_cast<jlong>(action_server);
+  return jaction_server;
+}
+
+#define RCLJAVA_ACTION_SERVER_TAKE_REQUEST(Type) \
```

Comment:
In this case, we'd be re-listing all parameters to the function where this macro is called. It seems a bit verbose and I'm not sure what is gained. If we want to be explicit about variables used, I'd rather just remove the macro and copy-paste the contents where we need them. But, this would be very verbose and I've opted to use a macro for brevity.

Reply:
the macro sounds good